utils/errors/errors.go

```diff
@@ -0,0 +1,114 @@
+package errors
+
+import (
+	"context"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	"github.com/gogo/protobuf/grpc"
+	"github.com/tendermint/tendermint/libs/log"
+	"golang.org/x/xerrors"
+	grpc2 "google.golang.org/grpc"
+)
+
+type causer interface {
+	Cause() error
+}
+
+// Is checks if the error is a (wrapped) error of the given type
+func Is[T error](err error) bool {
+	switch e := err.(type) {
+	case nil:
+		return false
+	case T:
+		return true
+	case causer:
+		return Is[T](e.Cause())
+	case xerrors.Wrapper:
+		return Is[T](e.Unwrap())
```

Comment:
Does golang have a default stack or recursion limit? I prefer doing this in a loop to not hit any limits due to recursion (to avoid a dos vector), since there is no tail call optimization.

Reply:
I doubt you will ever get into a case with 10 layers of nested errors, so we should be safe from that, and recursion is much cleaner than looping in this case

Reply:
I really don't like making that assumption, especially for a low level utility function not dealing with consensus directly. And looping is more memory-efficient anyway. While I like recursion in general, I don't like it in imperative languages due to these issues. Doesn't look too bad with a loop.

Reply:
So the recursion limit is not tight at all (min_stack is 2048, with an open upper bound), I was being a bit facetious about "no more than 10 layers". Unless you create a ring dependency you'll never get into this (and in that case a loop would help either). Also, the stdlib does unwrapping the same way. So I'd say we don't have to do preemptive optimization here